refactor(amf): re-aligned smfcontext structure parameters

ganeshg87 opened the pull request:
Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

smf_context_t structure parameters re-aligned, removed unused members


## Test Plan

- verified Registration, PDU Session Establishment and Release, Deregistration using UERANSIM
![image](https://user-images.githubusercontent.com/83060027/146008692-48995455-51d5-4666-ac95-e1675cf072ee.png)
[mme_reg_14_Dec_21_R1.log](https://github.com/magma/magma/files/7711977/mme_reg_14_Dec_21_R1.log)

- verified test oai
![image](https://user-images.githubusercontent.com/83060027/146008593-76d7d3c6-fbe7-4796-8a20-1b2ba5703aef.png)

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

ganeshg87 committed:
refactor(amf): realigned smfcontext parameters

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
+  uint8_t message_type;
+  uint8_t max_uplink;
+  uint8_t max_downlink;
+  uint32_t pdu_session_type;
```
Shouldn't this be an enum type?

ganeshg87 replied:
it was an Existing code.
- I think mainly to pass SMF GRPC message directly without converting to the corresponding enum.
- changing to Enum requires more amf changes for message_type, pdu_session_type, dl_ambr_unit.

@ssanadhya, Please suggest if needed we will take it as a separate activity.

ssanadhya replied:
@ganeshg87 , Pdu Session Type is specified as enum in [session_manager.proto](https://github.com/magma/magma/blob/master/lte/protos/session_manager.proto#L695) and the AMF GRPC servicer interprets it as [pdu_session_type_t](https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/grpc_service/AmfServiceImpl.cpp#L83), which is also an [enum](https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/include/n11_messages_types.h#L104). Am I missing something here?

ganeshg87 replied:
**session_manager**.proto :
// PduSessionType
enum PduSessionType {
   IPV4 = 0;
   IPV6 = 1;
   IPV4IPV6 = 2;
   UNSTRUCTURED = 3;
}
//nas message enum values as per spec(501)
enum class M5GPduSessionType : uint8_t {
  IPV4         = 0b001,
  IPV6         = 0b010,
  IPV4V6       = 0b011,
  UNSTRUCTURED = 0b100,
  ETHERNET     = 0b101,
};

both enum values are mis-match. both enums should follow the same values. 
If I replace enum defined in sesssion_manager.proto,  Need to validate sessiond & amf interworking behavior.

I accept your review comment, these all these should define as enum

ssanadhya replied:
@ganeshg87 , what's the plan to revise this PR?

ganeshg87 replied:
@ssanadhya, I have addressed review comments related to Enum change and updated all changes in current PR.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
 } smf_proc_data_t;
 
+typedef struct session_ambr_s {
+  uint8_t dl_ambr_unit;
```
Enum?

ganeshg87 replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
+typedef struct session_ambr_s {
+  uint8_t dl_ambr_unit;
+  uint16_t dl_session_ambr;
+  uint8_t ul_ambr_unit;
```
Enum?

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
-  SSCModeMsg ssc_mode;
+  uint8_t pdu_session_id;
+  uint8_t pti;
+  uint8_t message_type;
```
enum?

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
-  pdu_session_resource_setup_rsp_t pdu_resource_setup_rsp;
-  pdu_session_resource_to_release_list pdu_resource_release_req;
-  pdu_session_resource_to_release_list pdu_resource_release_rsp;
+  qos_flow_request_list_t subscribebed_qos_profile;
```
```suggestion
  qos_flow_request_list_t subscribed_qos_profile;
```

ganeshg87 replied:
addressed review comment

ganeshg87 replied:
addressed review comment

ganeshg87 committed:
Addressed review comments(R0)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 committed:
Addressed Review comments(R1)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
Hi @ssanadhya, 
Refactoring Existing Code will consume additional time and testing efforts. It will be difficult manage to  Stateless Feature deliverable timelines. We would need your support to raise separate tasks if changes will take more time.
cc: @panyogesh

ganeshg87 commented:
verified basic Registration, PDU Session Setup, and Release, Deregistration using UERANSIM.
[mme_reg_pdu_rel_derg_20.log](https://github.com/magma/magma/files/7906872/mme_reg_pdu_rel_derg_20.log)

![image](https://user-images.githubusercontent.com/83060027/150379398-0b521d8e-2e14-42f1-81ed-8ad1e1b28728.png)
![test_oai_20_jan](https://user-images.githubusercontent.com/83060027/150379521-2586b91b-d221-49b8-b4d6-1b784274f41a.png)

ganeshg87 commented:
@ssanadhya,
 I have addressed Enum-related changes, Please provide your feedback.

Thank you
Ganesh

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp:
```diff
@@ -307,10 +311,11 @@ int nas5g_message_encode(
     }
     OAILOG_INFO(
-        LOG_AMF_APP, "Encoded msg(nas5g) id: [%x]-name [%s]",
+        LOG_AMF_APP, "Encoded msg(nas5g) id: [%lu]-name [%s]",
```
this should be `%u`?

ganeshg87 replied:
@ardzoht, 

the actual intention was to print hex value for message_type.I have reverted back to hex print.

000465 Mon Jan 24 11:19:10 2022 7F76CE0E1700 INFO  AMF-AP tasks/amf/nas5g_message.cpp     :0327    Encoded msg(nas5g) id: [**56**]-name [AUTH_REQUEST].

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp:
```diff
@@ -320,9 +325,11 @@ int nas5g_message_encode(
 
     OAILOG_INFO(
-        LOG_AMF_APP, "Encoded msg(nas5g) id: [%x]-name [%s]",
+        LOG_AMF_APP, "Encoded msg(nas5g) id: [%lu]-name [%s]",
```
same comment here regarding `%u` formatting

ganeshg87 replied:
the actual intention was to print hex value for message_type.I have reverted back to hex print.

ardzoht reviewed (approved):
lgtm, left 2 minor comments, can be merged with @ssanadhya's approval

ganeshg87 committed:
Addressed review comments(R2)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:
```diff
@@ -175,59 +175,59 @@ static int amf_as_establish_req(amf_as_establish_t* msg, int* amf_cause) {
           msg->is_amf_ctx_new, *amf_cause, decode_status);
       break;
-    case M5G_SERVICE_REQUEST:  // SERVICE_REQUEST:
+    case M5GMessageType::M5G_SERVICE_REQUEST:  // SERVICE_REQUEST:
```
```suggestion
    case M5GMessageType::M5G_SERVICE_REQUEST: 
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -1056,15 +1065,17 @@ int construct_pdu_session_reject_dl_req(
-    msg->plain.amf.header.message_type = DLNASTRANSPORT;
-    msg->header.security_header_type   = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+    msg->plain.amf.header.message_type =
+        static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
What's the need to cast from "M5GMessageType" as "M5GMessageType"?

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp:
```diff
@@ -185,13 +185,17 @@ int nas5g_message_decode(
 
   OAILOG_INFO(
-      LOG_AMF_APP, "Decoded msg(nas5g) id: [%x]-name [%s]",
+      LOG_AMF_APP, "Decoded msg(nas5g) id: [%lu]-name [%s]",
```
Please address these lint comments.

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -707,12 +716,14 @@ TEST(test_dlnastransport, test_dlnastransport) {
   msg.header.extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.plain.amf.header.message_type          = DLNASTRANSPORT;
+  msg.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
What's the need to cast from "M5GMessageType" to "M5GMessageType"?

ganeshg87 replied:
addressed review comment, typecast is not needed

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -707,12 +716,14 @@ TEST(test_dlnastransport, test_dlnastransport) {
-  msg.header.message_type                    = DLNASTRANSPORT;
-  msg.header.sequence_number                 = 1;
+  msg.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Same comment as above for casting from M5GMessageType to M5GMessageType, please check in rest of the code as well.

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:
```diff
@@ -761,7 +759,8 @@ int amf_app_handle_pdu_session_accept(
       M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.security_protected.plain.amf.header.message_type = DLNASTRANSPORT;
+  msg.security_protected.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Casting "M5GMessageType" to "M5GMessageType"

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp:
```diff
@@ -160,7 +161,8 @@ int pdu_session_resource_release_request(
       M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.security_protected.plain.amf.header.message_type = DLNASTRANSPORT;
+  msg.security_protected.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Same comment as above

ganeshg87 replied:
addressed review comment

ssanadhya reviewed (approved):
Looks good, with minor nits. Please revise and then mark it as Ready2Merge. cc: @ardzoht

ganeshg87 committed:
Addressed Review comments(R3)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
> Looks good, with minor nits. Please revise and then mark it as Ready2Merge. cc: @ardzoht

@ssanadhya addressed review comments and marked Ready2Merge. cc: @ardzoht
